### PR TITLE
SPARK-1909 Green blob of user join/left messages now is on top of the transcript window.

### PR DESCRIPTION
**Pull request by Alameyo**
New joined/left and other presence messages still will appear durning conversation, but not informations about who joined/left before we joined the room.

### CODE REVIEW

**Comment by wrooot** [comment]
Suggest assigning to SPARK-1909 when merging.

**Comment by Alameyo** [comment]
So:
!e.getTimestamp().isBefore(drawTime) - exclude things that were send to chat before that time, i.e. messagesEntries as well as customEntries.
!(e instanceof CustomTextEntry) - exclude CustomTextEntres (joined/left but also changed subjects and maybe some others).
|| -  MessageEntries and other will always come through because they aren't CustomEntries. CustomEntries will come only if their timestamp is after joining the room (after drawTime ).

**Comment by guusdk** [comment]
Uff - could you at the very least document this in the code? I'll not remember this in two days...

**Comment by Alameyo** [comment]
Ok.

**Comment by Alameyo** [comment]
This piece of code is called everytime since, entries always are out of order, because Openfire sends first "archival" presence entries which we timestamp when they come and then archival MessageEntries which have delay extension. The only exception is when there are only presence entries entries in room. However I understand it might not be the case with every server or maybe even you can set this in Openfire.

**Comment by wrooot** [comment]
Is this still requiring changes?

**Comment by guusdk** [comment]
Yes, I feel that we need to change this. I really dislike this functionality to be added in a piece of code that ideally is never invoked. The code where it is put in was only added to correct some out-of-order scenario, that really should not occur in the first place.

**Comment by Alameyo** [comment]
Yes, I need to work on this a bit more.

**Comment by Alameyo** [comment]
Generally:
- now exist isDelayed() method for entries.
 - there is change in sorting in case of messages out of order. 
 - CustomEntries don't invoke sorting (since CostumMessages have no delay so that would messing things )
 - rest are mostly additional Constructors.

File: core/src/main/java/org/jivesoftware/spark/ui/TranscriptWindow.java [hunk 1]
```diff
@@ -122,11 +122,11 @@ protected synchronized void add( TranscriptWindowEntry entry )
                                                 // Clear and refill the UI component.
                                                 try
                                                 {
-                                                    entries.sort( Comparator.comparing( TranscriptWindowEntry::getTimestamp ) );
+                                                    entries.sort( Comparator.comparing(TranscriptWindowEntry::isDelayed).thenComparing( TranscriptWindowEntry::getTimestamp ) );
                                                     clear();
                                                     for ( TranscriptWindowEntry e : entries )
                                                     {
-                                                        e.addTo( transcriptWindow );
+                                                    	e.addTo( transcriptWindow );
```
Comment (guusdk) [inline core/src/main/java/org/jivesoftware/spark/ui/TranscriptWindow.java]: This is an unintentional whitespace change, I think.

File: core/src/main/java/org/jivesoftware/spark/ui/TranscriptWindowEntry.java [hunk 2]
```diff
@@ -89,4 +103,13 @@ public int hashCode()
     {
         return timestamp.toInstant().hashCode();
     }
+    /**
```
Comment (guusdk) [inline core/src/main/java/org/jivesoftware/spark/ui/TranscriptWindowEntry.java]: There's some whitespace oddity here too.

File: core/src/main/java/org/jivesoftware/spark/ui/TranscriptWindowEntry.java [hunk 3]
```diff
@@ -38,13 +39,26 @@
...
+    	this(timestamp, false);    	
+    }
+    /**
+     * Constructs a new entry.
+     *
+     * When displayed in context of a conversation, chat entries are ordered by time. This is why the time component
+     * of an entry is non-optional. Note that multiple entries can exist that have the same time component.
+     *
+     * @param timestamp The timestamp of the entry (cannot be null).
+     * @param isDelayed Set true if timestamp come from delay extension.
```
Comment (guusdk) [inline core/src/main/java/org/jivesoftware/spark/ui/TranscriptWindowEntry.java]: I'd change the description of the `isDelayed` argument, as it should be true for any message that is 'historic'. A 'delay' extension is just one way of expressing that. There might be other ways of expressing this too.

**Comment by guusdk** [comment]
Thank you!

**Pull request merged**